### PR TITLE
Fix some `Callable` false positives

### PR DESCRIPTION
**Pull request by RuRo**
## Describe your changes

This PR fixes some false positive errors with `Callable` types.

1) According to [PEP 484](https://www.python.org/dev/peps/pep-0484/#using-none) "When used in a type hint, the expression `None` is considered equivalent to `type(None)`." Currently, this is not the case for return values of `Callable` type hints.

    This leads to "funny" error messages like this:
    ```python
    attrs_strict._error.CallableError: Error with: call_me . Expected Callable signature: typing.Callable[[int, int], NoneType] got: 
    (a:int, b:int) -> None. None should be <class 'NoneType'>
    ```
    from
    ```python
    expected = Callable[[int, int], None]
    def call_me(a:int, b:int) -> None:
        pass
    ```

2) Functions with default values for arguments weren't handled properly.<br/>
    With this PR, the following function
    ```python
    def foo(a: int, b: int = 123) -> str: ...
    ```
    will also satisfy `Callable[[int], str]` and not just `Callable[[int, int], str]`.

3) Functions with keyword-only arguments weren't handled properly.<br/>
    With this PR,
    ```python
    def foo1(a: int, *, b: int = 123) -> str: ...
    def foo2(a: int, *, b: int) -> str: ...
    ```
     `foo1` will satisfy `Callable[[int], str]` but not `Callable[[int, int], str]`.
     `foo2` won't satisfy any of the above signatures (since `Callable` is positional-only).

## Testing performed

I added test cases for each item in "Describe your changes" (both false positives and false negatives).

### CODE REVIEW

**Comment by RuRo** [comment]
I am assuming, that you aren't pinning the exact dependency versions for CI? Sphinx version 4 came out fairly recently and the docs are now apparently failing. I didn't touch anything documentation-related, so I doubt that this is my fault.

**Comment by RuRo** [comment]
@erikseulean wdyt?

**Comment by RuRo** [comment]
I've also fixed a few more `Callable` bugs, PTAL.

**Comment by erikseulean** [comment]
Sorry, for some reasons I missed this completely. Apologise. Will review today.

**Review by erikseulean — COMMENTED** [review]
This LGTM I'll let @saytosid have another look before we can merge it.

**Comment by erikseulean** [comment]
@RuRo could you please squash your commits and rebase ?

File: src/attrs_strict/_type_validation.py
Comment (saytosid) [inline src/attrs_strict/_type_validation.py:226]: It is a bit concerning that we exclude support for such functions altogether. They are often used to improve readability in client code, we should allow them IMO.

I understand that `typing.Callable` being positional only is causing this problem, how do people feel about emitting a warning and letting it slide? Or better - provide a flag to control this "strictness"? 💭

**Comment by RuRo** [comment]
> @RuRo could you please squash your commits and rebase ?

I think the current changes are already rebased on top of the current master. If you can't squash the commits automatically when merging the PR for some reason, I can squash them after the current changes are approved.

File: src/attrs_strict/_type_validation.py
Comment (RuRo) [reply src/attrs_strict/_type_validation.py:226]: To be clear, there are essentially 6 kinds of arguments:
```python
def regular(a: int) -> None: pass
def regular_default(a: int = 0) -> None: pass

def positional_only(a: int, /) -> None: pass
def positional_only_default(a: int = 0, /) -> None: pass

def keyword_only(*, a: int) -> None: pass
def keyword_only_default(*, a: int = 0) -> None: pass
```
Although `typing.Callable` can only specify types for positional arguments, the only "unsupported" argument kind is `keyword_only`. If you think about it, out of these 6 functions, `keyword_only` is the only function, which you just can't call by passing only positional arguments. It is not clear to me, which `Callable` prototype are you suggesting should accept the `keyword_only` function. `Callable[[int], None]`? `Callable[[], None]`? Neither of those are correct, since `keyword_only(3)` and `keyword_only()` are both wrong.

I don't think, that I've seen non-defaulted keyword-only arguments that frequently in production code. And if someone really needs to accept such a `Callable`, they can always annotate the attribute with plain `Callable` without the generic specialization.

```python
@attr.s
class Any:
    any_function = attr.ib(type=typing.Callable, validator=[attrs_strict.type_validator()])

@attr.s
class Strict0:
    strict_function = attr.ib(type=typing.Callable[[], None], validator=[attrs_strict.type_validator()])

@attr.s
class Strict1:
    strict_function = attr.ib(type=typing.Callable[[int], None], validator=[attrs_strict.type_validator()])
```

Which gives the following results:

|                         | Any | Strict0 | Strict1 |
|------------------------:|:---:|:-------:|:-------:|
|                 regular |  🟢 |    ❌   |    🟢   |
|         regular_default |  🟢 |    🟢   |    🟢   |
|         positional_only |  🟢 |    ❌   |    🟢   |
| positional_only_default |  🟢 |    🟢   |    🟢   |
|            keyword_only |  🟢 |    ❌   |    ❌   |
|    keyword_only_default |  🟢 |    🟢   |    ❌   |

Comment (saytosid) [reply src/attrs_strict/_type_validation.py:226]: >  And if someone really needs to accept such a Callable, they can always annotate the attribute with plain Callable without the generic specialization.

I had forgotten about that, I can live with it 👍

**Review by saytosid — COMMENTED** [review]
LGTM. Thanks for the contribution! 🎉 

Can you please squash the test and implementation commits of each kind?

**Comment by RuRo** [comment]
Sure. Just to make sure, by "squash the test and implementation commits of each kind" do you mean

1) squash all the tests together and all the implementation commits together 
2) squash each implementation commit with its respective test commit
3) squash all the commits

?

**Comment by RuRo** [comment]
I've decided that you probably meant (2) and did that.

**Review by saytosid — APPROVED** [review]
Sorry for the delay. This LGTM :)

**Comment by RuRo** [comment]
It's that time of the week, where I go around all the PRs I am participating in and ping everyone!

@erikseulean @saytosid can I do anything to get this PR accepted?

**Review by gaborbernat — APPROVED** [review]

**Review by gaborbernat — CHANGES_REQUESTED** [review]
@RuRo if you fix the CI I'll approve.

**Comment by RuRo** [comment]
Oof. I haven't noticed, that you still support python 2.7. I'll fix the syntax errors.

As for the documentation failing, see my earlier comment. That problem wasn't caused by me, but by your unpinned sphinx version getting updated. You'll need to either pin the version or change the language setting.

**Comment by gaborbernat** [comment]
Change the language settings, feel free to add it here or even better a parallel pr 🙏

**Review by gaborbernat — APPROVED** [review]

**Pull request merged**
